File: Algorithms/p417_Pacific_Atlantic_Water_Flow_2.py
class Solution(object):
    def pacificAtlantic(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[List[int]]
        """
        if not matrix or not matrix[0]:
            return []
        m = len(matrix)
        n = len(matrix[0])
        p_table = [[False for j in range(n)] for i in range(m)]
        a_table = [[False for j in range(n)] for i in range(m)]
        q_p = collections.deque()
        q_a = collections.deque()
        for i in range(m):
            if not p_table[i][0]:
                p_table[i][0] = True
                q_p.append((i,0))
            if not a_table[i][n-1]:
                a_table[i][n-1] = True
                q_a.append((i,n-1))
        for j in range(n):
            if not p_table[0][j]:
                p_table[0][j] = True
                q_p.append((0,j))
            if not a_table[m-1][j]:
                a_table[m-1][j] = True
                q_a.append((m-1,j))
        self.flood(matrix,p_table,q_p,m,n)
        self.flood(matrix,a_table,q_a,m,n)
        res = []
        for i in range(m):
            for j in range(n):
                if p_table[i][j] and a_table[i][j]:
                    res.append([i,j])
        return res

    def flood(self, matrix, table, q, m, n):
        while q:
            i = q[0][0]
            j = q[0][1]
            q.popleft()
            if i>0 and not table[i-1][j] and matrix[i][j]<=matrix[i-1][j]:
                table[i-1][j] = True
                q.append((i-1,j))
            if i<m-1 and not table[i+1][j] and matrix[i][j]<=matrix[i+1][j]:
                table[i+1][j] = True
                q.append((i+1,j))
            if j>0 and not table[i][j-1] and matrix[i][j]<=matrix[i][j-1]:
                table[i][j-1] = True
                q.append((i,j-1))
            if j<n-1 and not table[i][j+1] and matrix[i][j]<=matrix[i][j+1]:
                table[i][j+1] = True
                q.append((i,j+1))

File: Algorithms/p417_Pacific_Atlantic_Water_Flow/p417_Pacific_Atlantic_Water_Flow_2.cpp
class Solution {
public:
    vector<pair<int, int>> pacificAtlantic(vector<vector<int>>& matrix) {
        if(matrix.empty()||matrix[0].empty()) return {};
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<bool>> p_table(m,vector<bool>(n,0));
        vector<vector<bool>> a_table(m,vector<bool>(n,0));
        queue<pair<int,int>> q_p;
        queue<pair<int,int>> q_a;
        for(int i=0;i<m;++i){
            if(!p_table[i][0]){
                p_table[i][0] = 1;
                q_p.emplace(i,0);
            }
            if(!a_table[i][n-1]){
                a_table[i][n-1] = 1;
                q_a.emplace(i,n-1);
            }
        }
        for(int j=0;j<n;++j){
            if(!p_table[0][j]){
                p_table[0][j] = 1;
                q_p.emplace(0,j);
            }
            if(!a_table[m-1][j]){
                a_table[m-1][j] = 1;
                q_a.emplace(m-1,j);
            }
        }
        flood(matrix,p_table,q_p,m,n);
        flood(matrix,a_table,q_a,m,n);
        vector<pair<int,int>> res;
        for(int i=0;i<m;++i){
            for(int j=0;j<n;++j){
                if(p_table[i][j]&&a_table[i][j]) res.emplace_back(i,j);
            }
        }
        return res;
    }
    void flood(vector<vector<int>>& matrix, vector<vector<bool>>& table, queue<pair<int,int>>& q, int m, int n) {
        while(!q.empty()){
            int i = q.front().first;
            int j = q.front().second;
            q.pop();
            if(i>0&&!table[i-1][j]&&matrix[i][j]<=matrix[i-1][j]){
                table[i-1][j] = 1;
                q.emplace(i-1,j);
            }
            if(i<m-1&&!table[i+1][j]&&matrix[i][j]<=matrix[i+1][j]){
                table[i+1][j] = 1;
                q.emplace(i+1,j);
            }
            if(j>0&&!table[i][j-1]&&matrix[i][j]<=matrix[i][j-1]){
                table[i][j-1] = 1;
                q.emplace(i,j-1);
            }
            if(j<n-1&&!table[i][j+1]&&matrix[i][j]<=matrix[i][j+1]){
                table[i][j+1] = 1;
                q.emplace(i,j+1);
            }
        }
    }
};
